src/openms/source/APPLICATIONS/TOPPBase.cpp

```diff
@@ -2382,6 +2412,57 @@ namespace OpenMS
     return true;
   }
 
+  bool TOPPBase::writeCWL_()
+  {
+    //store ini-file content in ini_file_str
+    QString out_dir_str = String(param_cmdline_.getValue("write_cwl").toString()).toQString();
+    if (out_dir_str == "")
+    {
+      out_dir_str = QDir::currentPath();
+    }
+    StringList type_list = ToolHandler::getTypes(tool_name_);
+    if (type_list.empty())
+      type_list.push_back(""); // no type for most tools (except GenericWrapper)
+
+    for (Size i = 0; i < type_list.size(); ++i)
+    {
+      QString write_cwl_file = out_dir_str + QDir::separator() + tool_name_.toQString() + type_list[i].toQString() + ".cwl";
+      outputFileWritable_(write_cwl_file, "write_cwl");
+
+      // set type on command line, so that getDefaultParameters_() does not fail (as it calls getSubSectionDefaults() of tool)
+      if (!type_list[i].empty())
+        param_cmdline_.setValue("type", type_list[i]);
+      Param default_params = getDefaultParameters_();
+
+      // add type to ini file
+      if (!type_list[i].empty())
+        default_params.setValue(this->ini_location_ + "type", type_list[i]);
+
+      std::stringstream ss;
+      ParamCWLFile paramFile;
+
+      std::string docurl = getDocumentationURL();
+      std::string category;
+      if (official_ || ToolHandler::getUtilList().count(tool_name_))
+      { // we can only get the docurl/category from registered/official tools
+        category = ToolHandler::getCategory(tool_name_);
+      }
+
+      std::vector<std::string> citation_dois;
+      citation_dois.reserve(citations_.size() + 1);
+      citation_dois.push_back(cite_openms_.doi);
+      for (auto& citation : citations_)
+      {
+        citation_dois.push_back(citation.doi);
+      }
+
+      paramFile.store(write_cwl_file.toStdString(), default_params,
+                      {version_, tool_name_, docurl, category, tool_description_, citation_dois});
+    }
+
+    return true;
+  }
+
```

Comment:
this looks like `writeCTD_`?!
Can you make this a template and just pass `ParamCWLFile` + `"write_cwl"` + `.cwl` ?!

Reply:
done